x/dualstaking/keeper/grpc_query_provider_delegators_test.go

```diff
@@ -0,0 +1,60 @@
+package keeper_test
+
+import (
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/testutil/common"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestProviderDelegatorsQuery(t *testing.T) {
+	ts := newTester(t)
+
+	ts.setupForDelegation(1, 1, 0, 0) // 1 delegator, 1 staked provider
+
+	_, delegator := ts.GetAccount(common.CONSUMER, 0)
+	_, provider := ts.GetAccount(common.PROVIDER, 0)
+
+	spec := ts.Spec("mock")
+
+	amountUint64 := uint64(100)
+	amount := sdk.NewCoin(epochstoragetypes.TokenDenom, sdk.NewIntFromUint64(amountUint64))
+
+	// delegate and query
+	_, err := ts.TxDualstakingDelegate(delegator, provider, spec.Index, amount)
+	require.Nil(t, err)
+
+	res, err := ts.QueryDualstakingProviderDelegators(provider)
+	require.Nil(t, err)
+	delegation := res.Delegations[0]
+	require.Equal(t, provider, delegation.Provider)
+	require.Equal(t, delegator, delegation.Delegator)
+	require.Equal(t, spec.Index, delegation.ChainID)
+	require.True(t, amount.Equal(delegation.Amount))
+
+	// partially unbond and query
+	unbondAmount := amount.Sub(sdk.NewCoin(epochstoragetypes.TokenDenom, sdk.OneInt()))
+	_, err = ts.TxDualstakingUnbond(delegator, provider, spec.Index, unbondAmount)
+	require.Nil(t, err)
+
+	bondedAmount := amount.Sub(unbondAmount)
+
+	res, err = ts.QueryDualstakingProviderDelegators(provider)
+	require.Nil(t, err)
+	delegation = res.Delegations[0]
+	require.Equal(t, provider, delegation.Provider)
+	require.Equal(t, delegator, delegation.Delegator)
+	require.Equal(t, spec.Index, delegation.ChainID)
+	require.True(t, bondedAmount.Equal(delegation.Amount))
+
+	// unbond completely and query (should not work)
+	ts.AdvanceEpoch() // TODO: if you put AdvanceBlock here, we get a critical error from unbonding. why?
```

Comment:
please share more details - what is the exact code to trigger the problem and what is the observed issue.

Reply:
already talked about it offline. For archive purposes, appending my original response for this issue:

The error originates from delegate.go:130. The full error print:

failed to decrease delegation ErrMsg: critical: delete delegation entry ErrMsg: not found -- [{Key:delegator Value:cosmos140605064sgka7lmc3e7p073cjw4jknfpvs2dq4} {Key:provider Value:cosmos1mm6fryf347vuzfwn7ge6w5a9lsmutxkpawdkmq} {Key:chainID Value:mockspec}]: not found -- [{Key:delegator Value:cosmos140605064sgka7lmc3e7p073cjw4jknfpvs2dq4} {Key:provider Value:cosmos1mm6fryf347vuzfwn7ge6w5a9lsmutxkpawdkmq} {Key:amount Value:1ulava}]: critical: delete delegation entry ErrMsg: not found -- [{Key:delegator Value:cosmos140605064sgka7lmc3e7p073cjw4jknfpvs2dq4} {Key:provider Value:cosmos1mm6fryf347vuzfwn7ge6w5a9lsmutxkpawdkmq} {Key:chainID Value:mockspec}]

The code flow that causes the error is as follows: delegate 100 coins, unbond 99 coins, advanceBlock, unbond 1 coin

For some reason, when you're trying to delete the entry after the second unbond (in which the delegation amount becomes zero), the entry cannot be found in delegationFS.

If instead of advanceBlock you put advanceEpoch, there is no error

Reply:
the problem was due to fixation entry bug with DelEntry() of a future entry that does not have a previous current entry. Fixed in #770.